### PR TITLE
Uart state period and optional forward

### PR DESCRIPTION
**Pull request by denniszollo**
still waiting on new SBP release but in the mean time,  @jacobmcnamee any time to review?

This computes the uart period and sends it out over the UART
It also makes a setting to decide whether to forward data from base out of the uart so that we can configure the other uarts the same as the FTDI uart when necessary or turn off the base station forwarding behavoir when it it saturating our serial line.

### CODE REVIEW

File: src/sbp.c [hunk 1]
```diff
@@ -348,26 +356,40 @@ void log_(u8 level, const char *msg, ...)
...
+  obs_period_ms = (now - last_obs_msg_ticks) / (double)CH_CFG_ST_FREQUENCY * 1000;
+  last_obs_msg_ticks = now;
   latency_accum_ms += (double) latency_ms;
+  period_accum_ms += (double) obs_period_ms;
   latency_count += 1;
 
   uart_state_msg.latency.current = (s32) ((LATENCY_SMOOTHING * ((float)latency_ms)) +
     ((1 - LATENCY_SMOOTHING) * (float) (uart_state_msg.latency.current)));
 
+  uart_state_msg.obs_period.current = (s32) ((LATENCY_SMOOTHING * ((float) (obs_period_ms)) +
```
Comment (jacobmcnamee) [inline src/sbp.c]: define a new smoothing constant or rename the existing one

File: src/sbp.c [hunk 2]
```diff
@@ -348,26 +356,40 @@ void log_(u8 level, const char *msg, ...)
 
 void log_obs_latency(float latency_ms)
 {
-  last_obs_msg_ticks = chVTGetSystemTime();
-
+  float obs_period_ms;
+  systime_t now = chVTGetSystemTime();
+  obs_period_ms = (now - last_obs_msg_ticks) / (double)CH_CFG_ST_FREQUENCY * 1000;
```
Comment (jacobmcnamee) [inline src/sbp.c]: combine with declaration?

File: src/sbp.c [hunk 3]
```diff
@@ -348,26 +356,40 @@ void log_(u8 level, const char *msg, ...)
...
     ((1 - LATENCY_SMOOTHING) * (float) (uart_state_msg.latency.current)));
 
+  uart_state_msg.obs_period.current = (s32) ((LATENCY_SMOOTHING * ((float) (obs_period_ms)) +
+    (1 - LATENCY_SMOOTHING) * (float) (uart_state_msg.obs_period.current)));
+
   /* Don't change the min and max latencies if we appear to have a zero latency
    * speed. */
-  if (latency_ms <= 0)
+  if (latency_ms <= 0 || obs_period_ms <= 0) {
+    log_warn("Incoherent observation reception: latency:%f, period: %f", latency_ms, obs_period_ms);
```
Comment (jacobmcnamee) [inline src/sbp.c]: spacing

File: src/sbp.c [hunk 4]
```diff
@@ -348,26 +356,40 @@ void log_(u8 level, const char *msg, ...)
...
   uart_state_msg.latency.current = (s32) ((LATENCY_SMOOTHING * ((float)latency_ms)) +
     ((1 - LATENCY_SMOOTHING) * (float) (uart_state_msg.latency.current)));
 
+  uart_state_msg.obs_period.current = (s32) ((LATENCY_SMOOTHING * ((float) (obs_period_ms)) +
+    (1 - LATENCY_SMOOTHING) * (float) (uart_state_msg.obs_period.current)));
+
   /* Don't change the min and max latencies if we appear to have a zero latency
    * speed. */
-  if (latency_ms <= 0)
+  if (latency_ms <= 0 || obs_period_ms <= 0) {
```
Comment (jacobmcnamee) [inline src/sbp.c]: `obs_period_ms` should always be positive since it is a u32 / (positive float)

File: src/sbp.c [hunk 5]
```diff
@@ -348,26 +356,40 @@ void log_(u8 level, const char *msg, ...)
...
-
+  }
   if (uart_state_msg.latency.lmin > latency_ms ||
       uart_state_msg.latency.lmin == 0) {
     uart_state_msg.latency.lmin = latency_ms;
   }
   if (uart_state_msg.latency.lmax < latency_ms) {
     uart_state_msg.latency.lmax = latency_ms;
   }
+  if (uart_state_msg.obs_period.pmin > (s32) obs_period_ms ||
```
Comment (jacobmcnamee) [inline src/sbp.c]: why are these cast to s32?

File: src/sbp.c (hunk 1 above)
Comment (denniszollo) [inline src/sbp.c]: done and will be refreshed on github shortly. Thanks for the review!

**Comment by denniszollo** [comment]
Responded to feedback and incorporated suggestions.  Should be ready to merge.

File: src/sbp.c (hunk 2 above)
Comment (denniszollo) [inline src/sbp.c]: Did originally, but then realized I had a bug when last_obs_msg_ticks was zero.  now I initialize to zero and only compute if i'm not in the first time through the loop.

File: src/sbp.c (hunk 4 above)
Comment (denniszollo) [inline src/sbp.c]: agreed

File: src/sbp.c (hunk 5 above)
Comment (denniszollo) [inline src/sbp.c]: me being silly.  No need to cast.

File: src/sbp.c (hunk 2 above)
Comment (jacobmcnamee) [inline src/sbp.c]: I just meant `float obs_period_ms = ...;`

**Comment by jacobmcnamee** [comment]
lgtm

**Pull request merged**